src/passes/TypeMerging.cpp

```diff
@@ -88,188 +91,331 @@ struct CastFinder : public PostWalker<CastFinder> {
 
   void visitCallIndirect(CallIndirect* curr) {
     if (!trapsNeverHappen) {
-      referredTypes.insert(curr->heapType);
+      castTypes.insert(curr->heapType);
     }
   }
 };
 
+// TODO: This is almost directly copied from type-updating.cpp. Deduplicate
+// this.
```

Comment:
Good idea - I think utilities for supertype-first and subtype-first sorting would be useful in a shared place.

Reply:
Do you mind if I do this as a follow-up?

Reply:
Sure, followup sgtm.